### PR TITLE
Prometheus: Remove running of duplicated metrics query

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
We are currently running metrics query `/api/v1/label/__name__/values?` when range changes (even if change is couple of milliseconds). This PR fixes this and only refreshes metrics when range changes, but we round up intervals to a minute.

**Master, running 3x the same requests:**
image

**This branch, not running the same queries over:**
image

Little downside could be, that is that if a user introduces a new metric, it is not going to appear instantly, but it is still going to be shown within 1 minute. 

**Which issue(s) this PR fixes**:

Related to https://github.com/grafana/grafana/issues/21893

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/components/PromQueryField.tsx [hunk 1]
```diff
@@ -188,7 +185,9 @@ class PromQueryField extends React.PureComponent<PromQueryFieldProps, PromQueryF
       });
     }
 
-    if (languageProvider !== prevProps.datasource.languageProvider || rangeChanged) {
+    const changedRangeToRefresh = this.rangeChangedToRefresh(range, prevProps.range);
+    // We want to rehresh metrics when language provider changes and/or when range changes (we round up intervals to a minute)
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/prometheus/components/PromQueryField.tsx:189]:
```suggestion
    // We want to refresh metrics when language provider changes and/or when range changes (we round up intervals to a minute)
```

File: public/app/plugins/datasource/prometheus/language_utils.ts [hunk 2]
```diff
@@ -185,3 +185,11 @@ export function fixSummariesMetadata(metadata: PromMetricsMetadata): PromMetrics
   }
   return { ...metadata, ...summaryMetadata };
 }
+
+export function roundMsToMin(miliseconds: number): number {
+  return roundSecToMin(miliseconds / 1000);
```
Comment (zoltanbedi) [inline public/app/plugins/datasource/prometheus/language_utils.ts:190]:
```suggestion
export function roundMsToMin(milliseconds: number): number {
  return roundSecToMin(milliseconds / 1000);
```

**Review by zoltanbedi — COMMENTED** [review]
This looks good. Would it make sense to add a test to PromQueryField?

**Comment by ivanahuckova** [comment]
> This looks good. Would it make sense to add a test to PromQueryField?

Yes it would, great suggestion. Also thanks for the spelling fixes. 💙

File: public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx [hunk 3]
```diff
@@ -140,7 +189,7 @@ describe('PromQueryField', () => {
     let cascader = screen.getByRole('button');
     // Should not show loading
     expect(cascader.textContent).toContain('Metrics');
-    checkMetricsInCascader(cascader, metrics);
+    checkMetricsInCascader(await screen.findByRole('button'), changedMetrics);
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:192]: @aocenas correct me if I am wrong, but I think this test should look like this cause we are checking if values changed.

**Review by zoltanbedi — APPROVED** [review]
Awesome! 🕶

File: public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx (hunk 3 above)
Comment (aocenas) [reply public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:192]: not sure exactly, I think the rerender on line 178 should be sync so on line 189 it should already get new DOM. Also the findByRole waits until it finds the element but I think the button should be already there anyway, it does not wait for it to change contents. Was there any issue that this solves?

File: public/app/plugins/datasource/prometheus/language_provider.ts [hunk 4]
```diff
@@ -443,8 +445,8 @@ export default class PromQlLanguageProvider extends LanguageProvider {
     // The rounding may seem strange but makes relative intervals like now-1h less prone to need separate request every
     // millisecond while still actually getting all the keys for the correct interval. This still can create problems
     // when user does not the newest values for a minute if already cached.
-    params.set('start', this.roundToMinutes(tRange['start']).toString());
-    params.set('end', this.roundToMinutes(tRange['end']).toString());
+    params.set('start', roundSecToMin(tRange['start']).toString());
+    params.set('end', roundSecToMin(tRange['end']).toString());
```
Comment (aocenas) [inline public/app/plugins/datasource/prometheus/language_provider.ts:449]: I am bit confused here. This code does not seem to be different which means we already did cache the query for 1 minute right? So the multiple queries should not happen even before even if there was a few second difference. Or is this different query from the one that was duplicated?

Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/language_provider.ts:449]: This is `fetchSeriesLabels` function (`/api/v1/series?...`) that caches the labels for specific metrics. So If you metric is for example `go_gc_duration_seconds` it caches all the label-values associated with this metric. 

What this PR does -> it makes sure that we are not fetching metrics too often. I have decided to solve this but not refreshing the metric if they are within the same 1 minute.

File: public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx (hunk 3 above)
Comment (ivanahuckova) [reply public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:192]: But this test is `'refreshes metrics when time range changes but dont show loading state'` but we've had the same metrics that we were comparing. Not the changed metrics.

Comment (aocenas) [reply public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:192]: I think the point is to still show the old metrics. So we don't actually want to wait for any change. Before when you rerendered with range it flashed the `loading` state until you got the new response. This basically checks that right after rerender we still show the old metrics without the loading state.

File: public/app/plugins/datasource/prometheus/language_provider.ts (hunk 4 above)
Comment (aocenas) [reply public/app/plugins/datasource/prometheus/language_provider.ts:449]: Ah ok so this is a different API call, makes sense then.

**Review by aocenas — APPROVED** [review]

File: public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx (hunk 3 above)
Comment (ivanahuckova) [inline public/app/plugins/datasource/prometheus/components/PromQueryField.test.tsx:192]:
```suggestion
    checkMetricsInCascader(cascader, metrics);
```

**Pull request merged**